shared/src/commonMain/kotlin/org/hyperskill/app/push_notifications/domain/interactor/PushNotificationsInteractor.kt

```diff
@@ -0,0 +1,78 @@
+package org.hyperskill.app.push_notifications.domain.interactor
+
+import org.hyperskill.app.auth.domain.interactor.AuthInteractor
+import org.hyperskill.app.core.domain.platform.Platform
+import org.hyperskill.app.devices.domain.model.Device
+import org.hyperskill.app.devices.domain.model.toDeviceType
+import org.hyperskill.app.devices.domain.repository.DevicesRepository
+import org.hyperskill.app.sentry.domain.interactor.SentryInteractor
+
+class PushNotificationsInteractor(
+    private val platform: Platform,
+    private val devicesRepository: DevicesRepository,
+    private val authInteractor: AuthInteractor,
+    private val sentryInteractor: SentryInteractor
+) {
+    suspend fun uploadFCMTokenToBackend(fcmToken: String): Result<Device> {
+        val isAuthorized = authInteractor.isAuthorized().getOrNull() ?: false
+        if (!isAuthorized) {
+            return Result.failure(Exception("User is not authorized"))
+        }
+
+        val currentFCMToken = getCurrentFCMToken()
+
+        return if (currentFCMToken != null && currentFCMToken != fcmToken) {
+            refreshFCMToken(oldToken = currentFCMToken, newToken = fcmToken)
+        } else {
+            activateFCMToken(fcmToken)
+        }.onFailure { sentryInteractor.captureErrorMessage("PushNotificationsInteractor: $it") }
+    }
+
+    internal fun handleUserDeauthorized() {
+        devicesRepository.clearCache()
+    }
+
+    internal suspend fun handleUserSignedOut() {
+        val currentFCMToken = getCurrentFCMToken()
+        devicesRepository.clearCache()
+
+        currentFCMToken?.let { disableFCMToken(it) }
```

Comment:
We never need to clear the cache in any case, because in both cases (token refresh failure and sign out) after the following sign in we can disable the old token if needed.